### PR TITLE
feat(artifact): add Walk method to artifact controller

### PR DESCRIPTION
**Pull request by heww**
Add Walk method to artifact controller which walks the artifact tree rooted at root, calling walkFn for each artifact in the tree, including root.

Signed-off-by: He Weiwei <hweiwei@vmware.com>

### CODE REVIEW

File: src/api/artifact/controller.go [hunk 1]
```diff
@@ -507,6 +512,48 @@ func (c *controller) RemoveLabel(ctx context.Context, artifactID int64, labelID
...
+		artifact := elem.Value.(*Artifact)
+		if err := walkFn(artifact); err != nil {
+			return err
+		}
+
+		if !artifact.HasChildren() {
+			continue
+		}
+
+		refs, err := c.artMgr.ListReferences(ctx, q.New(q.KeyWords{"parent_id": artifact.ID}))
```
Comment (ywk253100) [inline src/api/artifact/controller.go]: The `Artifact` model contains the references already, no need to list them again

File: src/api/artifact/controller.go [hunk 2]
```diff
@@ -507,6 +512,48 @@ func (c *controller) RemoveLabel(ctx context.Context, artifactID int64, labelID
...
+	for queue.Len() > 0 {
+		elem := queue.Front()
+		queue.Remove(elem)
+
+		artifact := elem.Value.(*Artifact)
+		if err := walkFn(artifact); err != nil {
+			return err
+		}
+
+		if !artifact.HasChildren() {
```
Comment (ywk253100) [inline src/api/artifact/controller.go]: Whether the artifact contains children or not can be determined by the length of references. The `HasChildren` isn't needed?

Comment (ywk253100) [reply src/api/artifact/controller.go]: I'm not sure it's the correct way to determine whether the artifact contains children by checking the media type. But the checking the references must be

Comment (heww) [reply src/api/artifact/controller.go]: I think we should check it by artifact media type, not the db query, and in artifact manager we should query references only when the media type is manifest index.

**Pull request merged**